tests/interfaces/test_batch_torch.py

```diff
@@ -0,0 +1,947 @@
+
+        res = execute(
+            [tape],
+            dev,
+            gradient_fn=param_shift,
+            gradient_kwargs={"shift": np.pi / 4},
+            interface="torch",
+        )[0]
+
+        res.backward()
+
+        for args in spy.call_args_list:
+            assert args[1]["shift"] == np.pi / 4
+
+    def test_unknown_gradient_fn_error(self):
+        """Test that an error is raised if an unknown gradient function
+        is passed"""
+        a = torch.tensor([0.1, 0.2], requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=1)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a[0], wires=0)
+            qml.RX(a[1], wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        res = execute([tape], dev, gradient_fn=lambda x: x, interface="torch")[0]
+
+        with pytest.raises(ValueError, match="Unknown gradient function"):
+            res.backward()
+
+    def test_incorrect_mode(self):
+        """Test that an error is raised if a gradient transform
+        is used with mode=forward"""
+        a = torch.tensor([0.1, 0.2], requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=1)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a[0], wires=0)
+            qml.RX(a[1], wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        with pytest.raises(
+            ValueError, match="Gradient transforms cannot be used with mode='forward'"
+        ):
+            execute([tape], dev, gradient_fn=param_shift, mode="forward", interface="torch")[0]
+
+    def test_forward_mode(self, mocker):
+        """Test that forward mode uses the `device.execute_and_gradients` pathway"""
+        dev = qml.device("default.qubit", wires=1)
+        spy = mocker.spy(dev, "execute_and_gradients")
+
+        a = torch.tensor([0.1, 0.2], requires_grad=True)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a[0], wires=0)
+            qml.RX(a[1], wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        res = execute(
+            [tape],
+            dev,
+            gradient_fn="device",
+            gradient_kwargs={"method": "adjoint_jacobian", "use_device_state": True},
+            interface="torch",
+        )[0]
+
+        # adjoint method only performs a single device execution, but gets both result and gradient
+        assert dev.num_executions == 1
+        spy.assert_called()
+
+    def test_backward_mode(self, mocker):
+        """Test that backward mode uses the `device.batch_execute` and `device.gradients` pathway"""
+        dev = qml.device("default.qubit", wires=1)
+        spy_execute = mocker.spy(qml.devices.DefaultQubit, "batch_execute")
+        spy_gradients = mocker.spy(qml.devices.DefaultQubit, "gradients")
+
+        a = torch.tensor([0.1, 0.2], requires_grad=True)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(a[0], wires=0)
+            qml.RX(a[1], wires=0)
+            qml.expval(qml.PauliZ(0))

```

Comment:
Some of the test cases might be worth being parametrized with `qml.var` and `qml.state`.

Reply:
I don't think parametrizing would be the best approach here 🤔 

- `qml.var` and `qml.state` are different logical pathways, and so should be different tests
- `qml.var` results in a different number of executions compared to `qml.expval`
- `qml.state` cannot be differentiated currently
- `qml.var` isn't supported by the adjoint method

Reply:
Oh makes sense! Then maybe a single case for `qml.var` could be helpful to have with `qml.tape.JacobianTape()` (the hessian is working well).